### PR TITLE
Verification of unchecked functions with bounds-safe interface

### PR DESCRIPTION
**Pull request by saeednj**
In this PR, we want to verify unchecked function against their bounds-safe interface. The changes are:

- Modify codegen to insert verification conditions at dynamic check points in the bit-code, to be picked up by Seahorn as a backend for verification.
- Added bounds for unchecked pointers, so we can have dynamic checks on their dereference.
- Added command line arguments for each of these two changes.
- Tests for codegen (standalone tests) and full verification pipeline (requires Seahorn to be installed and in path).
- Document: short description of seahorn and CheckedC+Seahorn pipeline

Note: The changes are not specific to Seahorn, any LLVM-based software verifier that uses the SV-COMP convention can be used as a backend.

See also #735

### CODE REVIEW

File: clang/lib/CodeGen/CGDynamicCheck.cpp [hunk 1]
```diff
@@ -380,6 +380,17 @@ BasicBlock *CodeGenFunction::EmitDynamicCheckFailedBlock() {
   // Add a "failed block", which will be inserted at the end of CurFn
   BasicBlock *FailBlock = createBasicBlock("_Dynamic_check.failed", CurFn);
   Builder.SetInsertPoint(FailBlock);
+  if (getLangOpts().InjectVerifierCalls) {
```
Comment (mgrang) [inline clang/lib/CodeGen/CGDynamicCheck.cpp]: `*` and `&` should align with the variable name, not the type name.

`T* x ==> T *x`
`T& x ==> T &x`

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# Verification of Unchecked codes with Seahorn
+
+Our goal is to verify whether unchecked functions violate their bounds-safe interface or not. We use [Seahorn](https://seahorn.github.io) tool as a backend verifier. Seahorn is a software verification framework. It can verify user-supplied assumption and assertions, as well as a number of built-in safety properties. We replace Seahorn's front-end with Checked-c clang to be able to process the bounds information and automatically insert necessary assertions to generate proper verification conditions.
```
Comment (dtarditi) [inline clang/docs/checkedc/verification-with-seahorn.md]: Typo: assumption->assumptions.

Also, we don't hyphenate Checked C.  It is "Checked C", not "Checked-C".

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    1. The function is not called with a non-NULL value for the pointers
+    2. The bounds lengths on argument pointers are non-negative
+* Function calls within the unckecked body:
+    1. Functions that are being called have bounds-safe interface as well
+
+For example, for this bounds information: `int *a : count(n)`, we will have: `assume(a != NULL); assume(n >= 0);`. The assumption on function calls having bounds-safe interface is needed to be able to have enough bounds information to propagate over the pointer assignments.
+
+## Installing Seahorn
+
+The following steps are quick listing for installing seahorn. For more detailed instructions please refer to Seahorn [github page](https://github.com/seahorn/seahorn).
```
Comment (dtarditi) [inline clang/docs/checkedc/verification-with-seahorn.md]: typo: seahorn->Seahorn

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+```bash
+INC= # /path/to/seahorn/build/run/include
+clang -c -emit-llvm -finject-verifier-calls -funchecked-pointers-dynamic-check -O0 -m64 -I $INC file.c -o file.bc
+sea pp file.bc -o file.pp
+sea ms file.pp -o file.ms
+sea opt file.ms -o file.opt
+sea horn --solve file.opt
+```
+
+The following code is a simple exmaple of an unchecked function with bounds-safe interface that we can verify with CheckedC+Seahorn.
```
Comment (dtarditi) [inline clang/docs/checkedc/verification-with-seahorn.md]: There is a typo in the spelling of example.

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    sassert(x_old + y_old == x + y);
+    return 0;
+}
+```
+
+
+## Checked C and Seahorn
+
+As mentioned above, we replace the front-end of Seahorn with Checked C, to use the bounds information and inject necessary assertion for the verification. The `sea fe` (Front End) is a shortcut for _clang_, _Pre processing_, _Mixed Semantics_ and _Compiler Optimization_ (`clang | pp | ms | opt`).
```
Comment (dtarditi) [inline clang/docs/checkedc/verification-with-seahorn.md]: Checked C -> the Checked C clang compiler.

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+Checked C will inject calls to verifier functions when `-finject-verifier-calls` is given. Note that this flag will generate a bitcode that does not link, because the calls do not have definitions. To enable adding bounds and consequently dynamic checks for unchecked pointers, `-funchecked-pointers-dynamic-check` flag should be given.
+
+We simply use Checked C clang instead of clang in the front-end pipeline to have the desired verification conditions. The following script shows an example pipeline. Each of the steps can be further configured with command line arguments.
+```bash
+INC= # /path/to/seahorn/build/run/include
+clang -c -emit-llvm -finject-verifier-calls -funchecked-pointers-dynamic-check -O0 -m64 -I $INC file.c -o file.bc
```
Comment (dtarditi) [inline clang/docs/checkedc/verification-with-seahorn.md]: Is there a simpler way to invoke Seahorn with the Checked C clang compiler as the front-end?  Or do people always need to use these 4 steps?

File: clang/docs/checkedc/verification-with-seahorn.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+To verify whether the unchecked functions violate their bounds-safe interface we follow these two steps:
+1. Add bounds to unchecked pointers to have dynamic checks at pointer dereferences
+2. Inject verifier sink functions (bad states) at dynamic check points
+
+We generate the LLVM bit code of the function that we are analyzing with the injected sink functions (`__VERIFIER_error`). We then pass the bit code to Seahorn. These are the main assumptions that we consider:
+* For each argument:
+    1. The function is not called with a non-NULL value for the pointers
+    2. The bounds lengths on argument pointers are non-negative
+* Function calls within the unckecked body:
```
Comment (kkjeer) [inline clang/docs/checkedc/verification-with-seahorn.md]: Typo: unckecked -> unchecked

File: clang/docs/checkedc/verification-with-seahorn.md (hunk 6 above)
Comment (saeednj) [reply clang/docs/checkedc/verification-with-seahorn.md]: I added a comment on how to only run `sea pf file.c`.

**Pull request merged**